packages/airnode-node/src/coordinator/calls/chain-limits.ts

```diff
@@ -0,0 +1,80 @@
+import flatMap from 'lodash/flatMap';
+import groupBy from 'lodash/groupBy';
+import map from 'lodash/map';
+import zipWith from 'lodash/zipWith';
+import * as logger from '../../logger';
+import { Config, ProviderStates, PendingLog, GroupedRequests } from '../../types';
+import { flattenRequests, groupRequests } from '../../requests/grouping';
+import { sortRequests } from '../../requests/sorting';
+
+function flattenAndSortRequests(grouped: GroupedRequests) {
+  return sortRequests(flattenRequests(grouped));
+}
+
+// Apply the chain request limit by dropping requests which exceed the limit. We do this mutably to increase
+// performance. This algorithm runs in O(P), where P = number of providers across all chains. When the chain limit is
+// exceeded, the latest requests of the largest provider shall be dropped.
+//
+// The algorithm works as following:
+// 1. Sort the providers by the number of actionable requests they reported
+// 2. Group the provider requests by the chain they operate on. Then for each chain do the following:
+//    1. Initialize available request limit (say A) to the chain limit specified in config.json
+//    2. Take the smallest unprocessed group of provider requests
+//    3. Take the maximum number of requests (say M) as you can. Let's say the current provider request count is R and
+//       there are N unprocessed providers (including the one currently processing). Then M = max(A / N, R). This is
+//       because taking M requests of the current provider implies taking M request from the others.
+//    4. Pick the oldest M request and discard other (extraneous) requests.
+//    5. Update A by subtracting M, go step 2.
+//
+// This algorithm naturally favours providers with smaller number of requests, while it also maximizes the number of
+// requests taken. E.g. N = 10, A = 56, R = 10. Then the algorithm picks (5,5,5,5,6,6,6,6,6). In general, if R > A / n,
+// then all providers end up with at least A / N request and last (A % N) end up with +1.
```

Comment:
And in the docs I plan to write something along these lines:
```
The "maximumConcurrency" specifies the maximum number of concurrent handler calls per single Airnode invocation. If you set this field to value X, then Airnode will guarantee that:
- At most X api calls are made to the API
- At most X transactions will be made by the blockchain providers of the respective chain

(This only applies to the requests initiated on chain. For example, requests triggered using http geteway are not included in this limit.)
```

Reply:
Here are the docs if you are interested: https://github.com/api3dao/api3-docs/pull/536

Reply:
Looks great, thanks for writing this up 👍🏻 

Is this right?

```suggestion
// then all providers end up with at least A / N request and the last provider (A % N) ends up with +1.
```

Reply:
no, this is meant to say that if there is 56 request limit and we have 10 providers with 15 requests each than we end up with the following:
5 5 5 5 6 6 6 6 6 6

A % N = 56 % 10 = 6 so 6 providers end up with +1 (the base amount is A / n = 5)